maven-core/src/main/java/org/apache/maven/project/MavenProject.java

```diff
@@ -1120,6 +1188,185 @@ private static String getProjectReferenceId(String groupId, String artifactId, S
         return buffer.toString();
     }
 
+    /**
+     * A List implementation that logs warnings when modified directly instead of using the proper add/remove methods.
+     * This is a wrapper that delegates all operations to the underlying list, so modifications to this list
+     * will affect the original list.
+     *
+     * @param <E> the type of elements in the list
+     * @since 3.9.10
+     */
+    private static class LoggingList<E> extends AbstractList<E> {
+        private static final String DISABLE_WARNINGS_PROPERTY = "maven.project.sourceRoots.warningsDisabled";
+        private final List<E> delegate;
+        private final String collectionName;
+
+        LoggingList(List<E> delegate, String collectionName) {
+            this.delegate = delegate;
+            this.collectionName = collectionName;
+        }
+
+        private void logWarning(String method) {
+            // Check if warnings are disabled
+            if (Boolean.getBoolean(DISABLE_WARNINGS_PROPERTY)) {
```

Comment:
Can we also check project properties here?
When we only check system properties - users will can only set it on cli ...

Reply:
Done